include/cantera/numerics/AdaptivePreconditioner.h

```diff
@@ -0,0 +1,177 @@
+          //!Function used to return compressed version of the matrix structure
+          virtual Eigen::SparseMatrix<double>* getMatrix();
+
+          //!Function used to set the dimensions of and construct the matrix structure - required for initialization and use of the class
+          //!@param nrows size_t number of rows in the structure
+          //!@param ncols size_t nubmer of columns in the structure
+          //!@param otherData void* for passing other data necessary for subclasses to initialize the matrix structure
+          virtual void setDimensions(size_t nrows,size_t ncols);
+
+          //!Function used to set a specific element of the matrix structure
+          //!@param row size_t specifying the row location
+          //!@param col size_t specifying the column location
+          //!@param element double value to be inserted into matrix structure
+          virtual void setElement(size_t row, size_t col, double element);//set element
+
+          //!Function used to set compressed version of the matrix structure
+          //!@param sparseMatrix a SUNMatrix pointer to a type of SUNMatrix
+          //!@param compress a bool dictating whether or not the set matrix needs compressed or not
+          virtual void setMatrix(Eigen::SparseMatrix<double>* sparseMatrix);  
+
+        /**
+         * 
+         * Reactor Level Functions
+         * 
+         **/
+
+        //!Function used to complete individual reactor setups
+          //!@param reactor A Reactor pointer
+          //!@param reactorStart an size_t providing the index location in which the state of the given reactor starts
+          virtual void reactorLevelSetup(Reactor* reactor, size_t reactorStart, double t, double* y, double* ydot, double* params);
+
+          //!Function used to complete individual reactor setups
+          //!@param reactor A IdealGasConstPressureReactor pointer
+          //!@param reactorStart an size_t providing the index location in which the state of the given reactor starts
+          virtual void reactorLevelSetup(IdealGasConstPressureReactor* reactor, size_t reactorStart, double t, double* y, double* ydot, double* params);
+
+          //!Function used to complete individual reactor setups
+          //!@param reactor A IdealGasConstPressureReactor pointer
+          //!@param reactorStart an size_t providing the index location in which the state of the given reactor starts
+          virtual void reactorLevelSetup(IdealGasReactor* reactor, size_t reactorStart, double t, double* y, double* ydot, double* params);
+
+          //!Function used to complete individual reactor setups
+          //!@param reactor A IdealGasConstPressureReactor pointer
+          //!@param reactorStart an size_t providing the index location in which the state of the given reactor starts
+          virtual void reactorLevelSetup(ConstPressureReactor* reactor, size_t reactorStart, double t, double* y, double* ydot, double* params);
+      };
+} 
+#endif
```

Comment:
Please make sure that all your files end with a single blank line. If you're using VS Code, you configure two settings to trim final newlines and end a file with the newline to make sure you have one.